### ISSUE TITLE
possible deadlock in com.spotify.feline.Feline with static initializer and 'allowances' Map

### ISSUE THREAD

**Issue by mattnworb** [spotify/futuristic-feline#18]
We have seen a few reports internally recently of what appears to be a deadlock within the Feline library, with thread dumps that look like the following:

the main thread is deep in executing the static initializer of `Feline` via a call to `com.spotify.feline.FelineMetricsRecorder.install()`:

```
"main" #1 prio=5 os_prio=0 cpu=4519.08ms elapsed=658.07s tid=0x00007f28a002b000 nid=0x7 waiting on condition  [0x00007f28a705b000]
   java.lang.Thread.State: RUNNABLE
	at java.lang.Class.forName0(java.base@11.0.15/Native Method)
	at java.lang.Class.forName(java.base@11.0.15/Class.java:398)
	at sun.reflect.generics.factory.CoreReflectionFactory.makeNamedType(java.base@11.0.15/CoreReflectionFactory.java:114)
	at ...
	at java.lang.Class.getGenericSuperclass(java.base@11.0.15/Class.java:949)
	at net.bytebuddy.description.type.TypeDescription$Generic$LazyProjection$ForLoadedSuperClass.resolve(TypeDescription.java:6570)
	at ...
	at net.bytebuddy.agent.builder.AgentBuilder$Default.doInstall(AgentBuilder.java:10916)
	at net.bytebuddy.agent.builder.AgentBuilder$Default.installOn(AgentBuilder.java:10842)
	at net.bytebuddy.agent.builder.AgentBuilder$Default$Delegator.installOn(AgentBuilder.java:12583)
	at com.spotify.feline.Feline.<clinit>(Feline.java:225)
	at com.spotify.feline.FelineMetricsRecorder.install(FelineMetricsRecorder.java:84)
	at com.spotify.feline.FelineMetricsRecorder.install(FelineMetricsRecorder.java:75)
```

and multiple other threads which were spawned from `main` prior to installing Feline are `in Object.wait()` with a lambda from `com.spotify.feline.Feline` at the top of the stack trace:

```
"New I/O worker #9" #33 daemon prio=5 os_prio=0 cpu=88.19ms elapsed=655.07s tid=0x00007f28a2bc1800 nid=0x32 in Object.wait()  [0x00007f27f9f6b000]
   java.lang.Thread.State: RUNNABLE
	at com.spotify.feline.Feline$$Lambda$363/0x00000008004ccc40.matches(Unknown Source)
	at net.bytebuddy.agent.builder.AgentBuilder$RawMatcher$ForElementMatchers.matches(AgentBuilder.java:1833)
	at net.bytebuddy.agent.builder.AgentBuilder$Default$ExecutingTransformer.doTransform(AgentBuilder.java:11870)
	at ...
	at net.bytebuddy.agent.builder.AgentBuilder$Default$ExecutingTransformer$ByteBuddy$ModuleSupport.transform(Unknown Source)
	at sun.instrument.TransformerManager.transform(java.instrument@11.0.15/TransformerManager.java:188)
	at sun.instrument.InstrumentationImpl.transform(java.instrument@11.0.15/InstrumentationImpl.java:563)
	at java.lang.ClassLoader.defineClass1(java.base@11.0.15/Native Method)
	at java.lang.ClassLoader.defineClass(java.base@11.0.15/ClassLoader.java:1017)
	at java.security.SecureClassLoader.defineClass(java.base@11.0.15/SecureClassLoader.java:174)
	at jdk.internal.loader.BuiltinClassLoader.defineClass(java.base@11.0.15/BuiltinClassLoader.java:800)
	at jdk.internal.loader.BuiltinClassLoader.findClassOnClassPathOrNull(java.base@11.0.15/BuiltinClassLoader.java:698)
	at jdk.internal.loader.BuiltinClassLoader.loadClassOrNull(java.base@11.0.15/BuiltinClassLoader.java:621)
	- locked <0x0000000724c99990> (a java.lang.Object)
	at jdk.internal.loader.BuiltinClassLoader.loadClass(java.base@11.0.15/BuiltinClassLoader.java:579)
	at jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(java.base@11.0.15/ClassLoaders.java:178)
	at java.lang.ClassLoader.loadClass(java.base@11.0.15/ClassLoader.java:522)
	...
```

```
"endpoint-change-notifier-9" #46 daemon prio=5 os_prio=0 cpu=9.60ms elapsed=654.94s tid=0x00007f28a2bf5800 nid=0x3f in Object.wait()  [0x00007f27f935f000]
   java.lang.Thread.State: RUNNABLE
	at com.spotify.feline.Feline$$Lambda$363/0x00000008004ccc40.matches(Unknown Source)
	at net.bytebuddy.agent.builder.AgentBuilder$RawMatcher$ForElementMatchers.matches(AgentBuilder.java:1833)
	at net.bytebuddy.agent.builder.AgentBuilder$Default$ExecutingTransformer.doTransform(AgentBuilder.java:11870)
	at ...
	at sun.instrument.InstrumentationImpl.transform(java.instrument@11.0.15/InstrumentationImpl.java:563)
	at java.lang.ClassLoader.defineClass1(java.base@11.0.15/Native Method)
	at java.lang.ClassLoader.defineClass(java.base@11.0.15/ClassLoader.java:1017)
	at java.security.SecureClassLoader.defineClass(java.base@11.0.15/SecureClassLoader.java:174)
	at jdk.internal.loader.BuiltinClassLoader.defineClass(java.base@11.0.15/BuiltinClassLoader.java:800)
	at jdk.internal.loader.BuiltinClassLoader.findClassOnClassPathOrNull(java.base@11.0.15/BuiltinClassLoader.java:698)
	at jdk.internal.loader.BuiltinClassLoader.loadClassOrNull(java.base@11.0.15/BuiltinClassLoader.java:621)
	- locked <0x00000007275497b0> (a java.lang.Object)
	at jdk.internal.loader.BuiltinClassLoader.loadClass(java.base@11.0.15/BuiltinClassLoader.java:579)
	at jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(java.base@11.0.15/ClassLoaders.java:178)
	...
```

It seems likely that the `com.spotify.feline.Feline$$Lambda$363` corresponds to the lambda in https://github.com/spotify/futuristic-feline/blob/e7de5ca34dd3a0a8f3843da2aeb08eac7cd50816/feline/src/main/java/com/spotify/feline/Feline.java#L216-L218 since that is the only lambda implementing the bytebuddy `ElementMatcher` interface.

Our best theory so far is that:

1. during application startup, a number of modules are started launching threads to do various things
1. a module which sets up `com.spotify.feline.FelineMetricsRecorder.install()` is executed, invoking the static initializer in `com.spotify.feline.Feline`
1.  while the `main` thread is executing Feline's static initializer which configures the bytebuddy AgentBuilder, other threads doing their work causes new classes to be loaded,  triggering the transformers in the bytebuddy agent - while it is in the process of being installed via the call to `AgentBuilder.installOn(instrumentation)` [here](https://github.com/spotify/futuristic-feline/blob/e7de5ca34dd3a0a8f3843da2aeb08eac7cd50816/feline/src/main/java/com/spotify/feline/Feline.java#L225) - and since the `ElementMatcher` instance we have passed into `AgentBuilder` for filtering types to transform involves executing a lambda defined in the `Feline` class, these threads are blocked waiting for the `Feline` class to finish being initialized 
   - this behavior fits in with the "initialization lock" mentioned in https://docs.oracle.com/javase/specs/jls/se11/html/jls-12.html#jls-12.4.1 (although it is a little curious that this is represented in the thread dump as the thread having a RUNNABLE state but being `in Object.wait()`)

What isn't totally clear at this point is why `main` / `com.spotify.feline.Feline<clinit>` is blocked in `java.lang.Class.forName0(java.base@11.0.15/Native Method)`, preventing the static initializer from finishing, which would unblock the other threads - or how to reliably reproduce the problem.

**Comment by mattnworb** [comment]
Here is the `javap` output to show what the classfile for `com.spotify.feline.Feline` looks like at version 1.1.1: https://gist.github.com/mattnworb/4dd3f7fd34089f4f701a9fb03c460596

I think what is relevant is the generated static method for the lambda `it -> allowances.containsKey(it.getName())`:

```
  private static boolean lambda$static$1(net.bytebuddy.description.type.TypeDescription);
    descriptor: (Lnet/bytebuddy/description/type/TypeDescription;)Z
    flags: (0x100a) ACC_PRIVATE, ACC_STATIC, ACC_SYNTHETIC
    Code:
      stack=2, locals=1, args_size=1
         0: getstatic     #8                  // Field allowances:Ljava/util/Map;
         3: aload_0
         4: invokeinterface #13,  1           // InterfaceMethod net/bytebuddy/description/type/TypeDescription.getName:()Ljava/lang/String;
         9: invokeinterface #14,  2           // InterfaceMethod java/util/Map.containsKey:(Ljava/lang/Object;)Z
        14: ireturn
      LineNumberTable:
        line 217: 0
      LocalVariableTable:
        Start  Length  Slot  Name   Signature
            0      15     0    it   Lnet/bytebuddy/description/type/TypeDescription;
```

and the code that calls it in the static initializer

```
  static {};
    descriptor: ()V
    flags: (0x0008) ACC_STATIC
    Code:
      stack=8, locals=2, args_size=0
         ...
       179: invokedynamic #62,  0             // InvokeDynamic #1:matches:()Lnet/bytebuddy/matcher/ElementMatcher;
       184: invokeinterface #60,  2           // InterfaceMethod net/bytebuddy/agent/builder/AgentBuilder.type:(Lnet/bytebuddy/matcher/ElementMatcher;)Lnet/bytebuddy/agent/builder/AgentBuilder$Identified$Narrowable;
       ...
```

I am not 100% sure how to read the bytecode for stuff like this but `#62` in the constant pool is

```
   #62 = InvokeDynamic      #1:#211       // #1:matches:()Lnet/bytebuddy/matcher/ElementMatcher;
...
  #211 = NameAndType        #321:#322     // matches:()Lnet/bytebuddy/matcher/ElementMatcher;
```

the generated static method for the lambda also appears in one of the bootstrapMethods for the class, which is where I think the actual class generation (i.e. the type that implements ElementMatcher) is done

```
BootstrapMethods:
...
  1: #139 REF_invokeStatic java/lang/invoke/LambdaMetafactory.metafactory:(Ljava/lang/invoke/MethodHandles$Lookup;Ljava/lang/String;Ljava/lang/invoke/MethodType;Ljava/lang/invoke/MethodType;Ljava/lang/invoke/MethodHandle;Ljava/lang/invoke/MethodType;)Ljava/lang/invoke/CallSite;
    Method arguments:
      #208 (Ljava/lang/Object;)Z
      #209 REF_invokeStatic com/spotify/feline/Feline.lambda$static$1:(Lnet/bytebuddy/description/type/TypeDescription;)Z
      #210 (Lnet/bytebuddy/description/type/TypeDescription;)Z
```

**Comment by mattnworb** [comment]
oops, did not mean to mark this as closed yet

**Comment by mattnworb** [comment]
I've released `1.1.2` with a possible fix for this (from #19), it should be on its way to Maven Central

**Comment by mattnworb** [comment]
Closing as we are feeling semi confident #19 addressed the issue. I'll reopen this if I'm wrong!
